rust/stub/src/pe_loader.rs

```diff
@@ -0,0 +1,91 @@
+use goblin::pe::PE;
+use uefi::{
+    prelude::BootServices,
+    table::{
+        boot::{AllocateType, MemoryType},
+        Boot, SystemTable,
+    },
+    Handle, Status,
+};
+
+pub struct Image {
+    pub base: *const u8,
+    pub length: usize,
+    pub entry: extern "efiapi" fn(Handle, SystemTable<Boot>) -> Status,
+}
+
+#[cfg(target_arch = "x86_64")]
+fn flush_instruction_cache(_start: *const u8, _length: usize) {
+    // x86_64 mandates coherent instruction cache
+}
+
+/// Loads and relocates a trusted PE file, returning a structure containing its allocated virtual memory range and entry point.
+pub unsafe fn load(boot_services: &BootServices, file_data: &[u8]) -> uefi::Result<Image> {
+    let pe = PE::parse(file_data).expect("Malformed PE file");
+
+    // Allocate all memory the image will need in virtual memory.
+    // We follow shim here and allocate as EfiLoaderCode.
```

Comment:
Do you have a pointer to documentation?

Reply:
It's documented in the UEFI spec (e.g. section 2.1.1) that EFI applications get their code loaded as EfiLoaderCode and their data as EfiLoaderData. However, we can't do that since they need to be contiguous. So I checked what shim does in its code, and it turns out it uses EfiLoaderCode for everything.

Reply:
Ok.